### PR TITLE
Enable extra commit properties with metadata delete

### PR DESCRIPTION
**Pull request by CodingCat**
due to the missing of setting commit msg, deletion finished with pure metadata operation cannot successfully adding snapshot summary specified by user via `org.apache.iceberg.spark.CommitMetadata`. This PR adds the commit msg reading for Spark 3.2/3.3/3.4 and also add the docs about how to use CommitMetadata

### CODE REVIEW

**Comment by CodingCat** [comment]
cc: @rdblue @RussellSpitzer @flyrain would you mind giving it a review? thank you!

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java [hunk 1]
```diff
@@ -448,4 +451,52 @@ public void testExtraSnapshotMetadataWithSQL() throws InterruptedException, IOEx
...
+    writerThread.join();
+    Set<String> threadNames = Sets.newHashSet();
+    spark.sql("SELECT * from " + tableName).show();
+    Table table = validationCatalog.loadTable(tableIdent);
+    for (Snapshot snapshot : table.snapshots()) {
+      threadNames.add(snapshot.summary().get("writer-thread"));
+    }
+    for (String t : threadNames) {
+      System.out.println(t);
+    }
```
Comment (flyrain) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:497]: Is it necessary?

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java [hunk 2]
```diff
@@ -448,4 +451,52 @@ public void testExtraSnapshotMetadataWithSQL() throws InterruptedException, IOEx
...
+    Dataset<Row> originalDf = spark.createDataFrame(expectedRecords, SimpleRecord.class);
+    originalDf.repartition(5, new Column("data")).select("id", "data").writeTo(tableName).append();
+    spark.sql("SELECT * from " + tableName + ".files").show();
+    System.out.println(
+        spark
+            .sql("EXPLAIN DELETE FROM " + tableName + " where id = 1")
+            .collectAsList()
+            .get(0)
+            .get(0));
+    System.out.println("finished inserting");
```
Comment (flyrain) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:472]: Can we remove `System.out.println`?

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java [hunk 3]
```diff
@@ -448,4 +451,52 @@ public void testExtraSnapshotMetadataWithSQL() throws InterruptedException, IOEx
...
+  public void testExtraSnapshotMetadataWithDelete()
+      throws InterruptedException, IOException, NoSuchTableException {
+    spark.sessionState().conf().setConfString("spark.sql.shuffle.partitions", "1");
+    sql("CREATE TABLE %s (id INT, data STRING) USING iceberg", tableName);
+    List<SimpleRecord> expectedRecords =
+        Lists.newArrayList(
+            new SimpleRecord(1, "a"), new SimpleRecord(2, "b"), new SimpleRecord(3, "c"));
+    Dataset<Row> originalDf = spark.createDataFrame(expectedRecords, SimpleRecord.class);
+    originalDf.repartition(5, new Column("data")).select("id", "data").writeTo(tableName).append();
+    spark.sql("SELECT * from " + tableName + ".files").show();
```
Comment (flyrain) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:465]: Do we need this line?

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala [hunk 4]
```diff
@@ -122,6 +122,8 @@ case class ExtendedDataSourceV2Strategy(spark: SparkSession) extends Strategy wi
         }
         filter
       }.toArray
+      // scalastyle:off
+      println("DeleteFromIcebergTable")
```
Comment (flyrain) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:126]: Is this change relevant?

File: docs/spark-configuration.md [hunk 5]
```diff
@@ -194,3 +194,20 @@ df.write
...
+Map<String, String> properties = Maps.newHashMap();
+properties.put("writer-thread", String.valueOf(Thread.currentThread().getName()));
+CommitMetadata.withCommitProperties(properties,
+        () -> {
+        spark.sql("DELETE FROM " + tableName + " where id = 1");
+        return 0;
+        },
+        RuntimeException.class);
+
+```
```
Comment (flyrain) [inline docs/spark-configuration.md:213]: Can we split this PR and move the doc change to a new PR?

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java [hunk 6]
```diff
@@ -445,4 +448,40 @@ public void testExtraSnapshotMetadataWithSQL() throws InterruptedException, IOEx
...
+                  properties,
+                  () -> {
+                    spark.sql("DELETE FROM " + tableName + " where id = 1");
+                    return 0;
+                  },
+                  RuntimeException.class);
+            });
+    writerThread.setName("test-extra-commit-message-delete-thread");
+    writerThread.start();
+    writerThread.join();
```
Comment (flyrain) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:477]: Nit: an empty line to make it more readable?

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java [hunk 7]
```diff
@@ -445,4 +448,40 @@ public void testExtraSnapshotMetadataWithSQL() throws InterruptedException, IOEx
...
+    Set<String> threadNames = Sets.newHashSet();
+    Table table = validationCatalog.loadTable(tableIdent);
+    for (Snapshot snapshot : table.snapshots()) {
+      threadNames.add(snapshot.summary().get("writer-thread"));
+    }
+    Assert.assertEquals(2, threadNames.size());
+    Assert.assertTrue(threadNames.contains(null));
+    Assert.assertTrue(threadNames.contains("test-extra-commit-message-delete-thread"));
```
Comment (flyrain) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:485]: Would this be a bit more expressive to show which snapshot has the property?
```
    List<Snapshot> snapshots = Lists.newArrayList(table.snapshots());
    Assert.assertEquals(2, snapshots.size());
    Assert.assertNull(snapshots.get(0).summary().get("writer-thread"));
    Assert.assertEquals("test-extra-commit-message-delete-thread", snapshots.get(1).summary().get("writer-thread"));
```

**Review by flyrain — COMMENTED** [review]
Thanks @CodingCat for the PR. Looks good to me overall. Left minor comments. I don't have much context with `CommitMetadata`. My understanding is that every time we generate a new snapshot. We will need to inject it if it is there. Is it possible to have the functionality of setting it in a base class of all snapshot producer?

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java [hunk 8]
```diff
@@ -445,4 +448,40 @@ public void testExtraSnapshotMetadataWithSQL() throws InterruptedException, IOEx
     Assert.assertTrue(threadNames.contains(null));
     Assert.assertTrue(threadNames.contains("test-extra-commit-message-writer-thread"));
   }
+
+  @Test
+  public void testExtraSnapshotMetadataWithDelete()
+      throws InterruptedException, IOException, NoSuchTableException {
```
[the remark this answers is not in the document]
Comment (CodingCat) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:454]: good point, updated

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 9]
```diff
@@ -157,6 +157,7 @@ public DeltaWriterFactory createBatchWriterFactory(PhysicalWriteInfo info) {
 
     @Override
     public void commit(WriterCommitMessage[] messages) {
+
```
[the remark this answers is not in the document]
Comment (CodingCat) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:160]: updated

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java (hunk 1 above)
Comment (CodingCat) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:497]: oops, forgot to cleanup

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java (hunk 3 above)
Comment (CodingCat) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:465]: removed

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java (hunk 2 above)
Comment (CodingCat) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:472]: removed

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java (hunk 6 above)
Comment (CodingCat) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:477]: updated

File: docs/spark-configuration.md (hunk 5 above)
Comment (CodingCat) [reply docs/spark-configuration.md:213]: updated

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java (hunk 7 above)
Comment (CodingCat) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java]: updated

**Comment by CodingCat** [comment]
> Thanks @CodingCat for the PR. Looks good to me overall. Left minor comments. I don't have much context with `CommitMetadata`. My understanding is that every time we generate a new snapshot. We will need to inject it if it is there. Is it possible to have the functionality of setting it in a base class of all snapshot producer?

thanks for the review @flyrain !

IIUC, you mean we move the code consuming CommitMetadata to a base class so that we will not fall into the situation like this PR tries to address?

 I thought about it when started working on this PR , it may involve a bit refactoring for iceberg-spark ...


in the current implementations,  `commitOperation` in [SparkWrite](https://github.com/apache/iceberg/blob/82880be39654bb94aaf370338bdd61f706caa6da/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java#L192)  is different with that in [SparkPositionDeltaWrite](https://github.com/apache/iceberg/blob/82880be39654bb94aaf370338bdd61f706caa6da/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java#L259) are pretty much the same ,  we can potentially move them to a base trait, my only concern is 

(1) do we want to do in this PR?
(2) in future, can we guarantee to use the same commit implementation for all snapshot producers (at least in Spark)?

File: docs/spark-configuration.md [hunk 10]
```diff
@@ -194,3 +194,4 @@ df.write
 | check-ordering       | true        | Checks if input schema and table schema are same  |
 | isolation-level | null | Desired isolation level for Dataframe overwrite operations.  `null` => no checks (for idempotent writes), `serializable` => check for concurrent inserts or deletes in destination partitions, `snapshot` => checks for concurrent deletes in destination partitions. |
 | validate-from-snapshot-id | null | If isolation level is set, id of base snapshot from which to check concurrent write conflicts into a table. Should be the snapshot before any reads from the table. Can be obtained via [Table API](../../api#table-metadata) or [Snapshots table](../spark-queries#snapshots). If null, the table's oldest known snapshot is used. |
+
```
Comment (flyrain) [inline docs/spark-configuration.md:197]: Can we remove this change?

**Review by flyrain — APPROVED** [review]
> IIUC, you mean we move the code consuming CommitMetadata to a base class so that we will not fall into the situation like this PR tries to address?

Exactly. We can refactor later. I'm OK with a quick fix like this.

>(2) in future, can we guarantee to use the same commit implementation for all snapshot producers (at least in Spark)?

I'm not sure if I got you correctly. I assume we do need the same implementation. Any concern?

**Comment by CodingCat** [comment]
> > IIUC, you mean we move the code consuming CommitMetadata to a base class so that we will not fall into the situation like this PR tries to address?
> 
> Exactly. We can refactor later. I'm OK with a quick fix like this.
> 
> > (2) in future, can we guarantee to use the same commit implementation for all snapshot producers (at least in Spark)?
> 
> I'm not sure if I got you correctly. I assume we do need the same implementation. Any concern?

what i mean was that if we move `commitOperation` to a base class, can we guarantee we always call the base class's commitOperation in future? (probably hard to make such a commitment, but that would break the effort of refactoring again)

**Comment by flyrain** [comment]
Merged. Thanks @CodingCat for the change. 

> what i mean was that if we move commitOperation to a base class, can we guarantee we always call the base class's commitOperation in future? (probably hard to make such a commitment, but that would break the effort of refactoring again)

Not sure how it looks finally, but worth to have a try.

**Pull request merged**